controllers/httpproxy_controller.go

```diff
@@ -0,0 +1,190 @@
+package controllers
+
+import (
+	"context"
+	"net"
+
+	"github.com/go-logr/logr"
+	certmanagerv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	projectcontourv1 "github.com/projectcontour/contour/apis/projectcontour/v1"
+)
+
+// HTTPProxyReconciler reconciles a HttpProxy object
+type HTTPProxyReconciler struct {
+	client.Client
+	Log               logr.Logger
+	Scheme            *runtime.Scheme
+	ServiceKey        client.ObjectKey
+	IssuerKey         client.ObjectKey
+	Prefix            string
+	DefaultIssuerName string
+	DefaultIssuerKind string
+	CreateDNSEndpoint bool
+	CreateCertificate bool
+}
+
+// +kubebuilder:rbac:groups=projectcontour.io,resources=httpproxies,verbs=get;list;watch
+// +kubebuilder:rbac:groups=projectcontour.io,resources=httpproxies/status,verbs=get
+// +kubebuilder:rbac:groups=externaldns.k8s.io,resources=dnsendpoints,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cert-manager.io,resources=certificates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=services/status,verbs=get
+
+// Reconcile creates/updates CRDs from given HTTPProxy
+func (r *HTTPProxyReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("httpproxy", req.NamespacedName)
+
+	// Get IngressRoute
+	hp := new(projectcontourv1.HTTPProxy)
+	objKey := client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}
+	err := r.Get(ctx, objKey, hp)
+	if k8serrors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		log.Error(err, "unable to get HTTPProxy resources")
+		return ctrl.Result{}, err
+	}
+
+	if hp.Annotations[excludeAnnotation] == "true" {
+		return ctrl.Result{}, nil
+	}
+
+	err = r.reconcileDNSEndpoint(ctx, hp, log)
+	if err != nil {
+		log.Error(err, "unable to reconcile DNSEndpoint")
+		return ctrl.Result{}, err
+	}
+
+	err = r.reconcileCertificate(ctx, hp, log)
+	if err != nil {
+		log.Error(err, "unable to reconcile Certificate")
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager initializes controller manager
+func (r *HTTPProxyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&projectcontourv1.HTTPProxy{}).
```

Comment:
The controller owns DNSEndpoint and Certificate so should be watched by
[`Owns`](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/builder#Builder.Owns) to kick reconcilation immediately when these resources are manually removed
or things like that.